AWSCore/Authentication/AWSSignature.m

```diff
@@ -370,77 +371,102 @@ - (NSString *)signRequestV4:(NSMutableURLRequest *)request
+                               signSessionToken:(BOOL)signSessionToken {
+
     return [[credentialsProvider credentials] continueWithSuccessBlock:^id _Nullable(AWSTask<AWSCredentials *> * _Nonnull task) {
+        // No nullability specifier on `credentials` return value means that we need to check for nil here to be safe
+        if (!task.result) {
+            NSString *description = @"Credentials result unexpectedly nil generating presigned URL";
+            NSDictionary *userInfo = @{
+                                       NSLocalizedDescriptionKey: description
+                                       };
+            NSError *error = [NSError errorWithDomain:AWSCognitoCredentialsProviderErrorDomain
+                                                 code:AWSCognitoCredentialsProviderErrorUnknown
+                                             userInfo:userInfo];
+            return [AWSTask taskWithError:error];
+        }
+
         AWSCredentials *credentials = task.result;
-        
-        //Implementation of V4 signaure http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
-        NSMutableString *queryString = [NSMutableString new];
-        
+
+        // Deconstruct the incoming URL into components for easier manipulation and inspection of individual pieces.
+        // We'll use the mutated components at the end of this method to construct the signed URL
+        NSURLComponents *urlComponents = [[NSURLComponents alloc] initWithURL:request.URL
+                                                      resolvingAgainstBaseURL:NO];
+
+        // Implementation of V4 signature http://docs.aws.amazon.com/AmazonS3/latest/API/sigv4-query-string-auth.html
+        // Start with existing query string parameters; signature parameters will be appended to them
+        NSMutableArray<NSURLQueryItem *> *queryItems = [[NSMutableArray alloc] initWithArray:urlComponents.queryItems];
+
         //Append Identifies the version of AWS Signature and the algorithm that you used to calculate the signature.
-        [queryString appendFormat:@"%@=%@&",@"X-Amz-Algorithm",AWSSignatureV4Algorithm];
-        
-        //Get ClockSkew Fixed Date
-        NSDate *currentDate = [NSDate aws_clockSkewFixedDate];
-        
-        //Format of X-Amz-Credential : <your-access-key-id>/<date>/<AWS-region>/<AWS-service>/aws4_request.
-        NSString *scope = [NSString stringWithFormat:@"%@/%@/%@/%@",
-                           [currentDate aws_stringValue:AWSDateShortDateFormat1],
-                           endpoint.regionName,
-                           endpoint.serviceName,
-                           AWSSignatureV4Terminator];
-        
-        NSString *signingCredentials = [NSString stringWithFormat:@"%@/%@",credentials.accessKey, scope];
-        //need to replace "/" with "%2F"
-        NSString *xAmzCredentialString = [signingCredentials stringByReplacingOccurrencesOfString:@"/" withString:@"\%2F"];
```

Comment:
This line is removed in the change, will this be handled by NSURLQueryItem?

Reply:
Yes, percent encoding is done at the point of conversion into a canonical request